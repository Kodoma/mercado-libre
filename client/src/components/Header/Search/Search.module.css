.autocompleteContainer {
    color: #333;
    font-size: 16px;
    width: 90%;
    margin: 0 auto;
    font-family: inherit;
    display: inline-block;
    right: 0px;
    position: absolute;
}

.autocomplete {
    width: 80%;
    display: inline-block;
    marginTop: 16px;
    margin: 0 10%;
}

.input {
    background: #fff;
    background: -moz-linear-gradient(top,#fff 0,#f1f1f1 100%);
    background: -ms-linear-gradient(top,#fff 0,#f1f1f1 100%);
    background: -o-linear-gradient(top,#fff 0,#f1f1f1 100%);
    background: -webkit-gradient(linear,left top,left bottom,color-stop(0%,#fff),color-stop(100%,#f1f1f1));
    background: -webkit-linear-gradient(top,#fff 0,#f1f1f1 100%);
    background: linear-gradient(to bottom,#fff 0,#f1f1f1 100%);
}

.searchButton {
    background: #EEEEEE !important;
    border: 1px solid #CDCCCB;
    color: #666;
    cursor: pointer;
    filter: progid:DXImageTransform.Microsoft.gradient(startColorstr='#EEEEEE',endColorstr='#EEEEEE',GradientType=0);
    height: 55px;
    position: absolute;
    width: 55px;
    top: 16px;
    left: 5%;
}

.searchIcon {
    position: absolute;
    left: 0;
    top: 0;
    background-position: 18px 18px;
    width: 44px;
    height: 40px;
    background-image: url(http://static.mlstatic.com/org-img/commons/assets/sprite.png?v1);
    background-repeat: no-repeat;
    display: inline-block;
    vertical-align: middle;
    text-indent: -9999px;
}

@media (max-width: 812px) {
    .autocompleteContainer {
        height: 48px;
    }

    .searchButton {
        display: none;
        visibility: hidden;
    }

    .autocomplete {
        width: 90%;
        display: inline-block;
        marginTop: 8px;
        marginLeft: 25%
    }
}