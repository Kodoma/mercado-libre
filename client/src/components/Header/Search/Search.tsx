import fetch from "cross-fetch";
import React from "react";
import TextField from "@material-ui/core/TextField";
import Autocomplete from "@material-ui/lab/Autocomplete";
import CircularProgress from "@material-ui/core/CircularProgress";
import styles from "./Search.module.css";
import { useForm } from "react-hook-form";
import { useRouter } from "next/router";
import { Button } from "@material-ui/core";
import {Result} from "../../../models/search";

const SearchAutocomplete = () => {
    const [open, setOpen] = React.useState(false);
    const [options, setOptions] = React.useState(ResultInitState);
    const [query, setQuery] = React.useState("");
    const [inputValue, setInputValue] = React.useState("");
    const loading = open && options.length === 0;
    const { handleSubmit } = useForm();
    const router = useRouter()

    const onSubmit = () => {
        router.push("/items?search="+decodeURI(query));
    };

    const onChangeHandle = async (value: string) => {
        if ( value.length >= 3) {
            setInputValue(value);
            const request = await fetch(
                "https://api.mercadolibre.com/sites/MLA/search?q="+encodeURI(value)
            );
            const response = await request.json();
            setOptions(response.results);
        }
    };

    React.useEffect(() => {
        if (!open) {
            setOptions([]);
        }
    }, [open]);

    return (
        <form onSubmit={handleSubmit(onSubmit)} className={styles.autocompleteContainer}>
            <Autocomplete
                id="q"
                style={{ width: '70%', display: 'inline-block', marginTop: '8px', left: '7.5%', position: 'relative'}}
                freeSolo
                open={open}
                onOpen={(ev) => {
                    setOpen(options.length >= 2);
                }}
                onClose={() => {
                    setOpen(false);
                }}
                onInputChange={(event, newInputValue) => {
                    setQuery(newInputValue);
                }}
                getOptionSelected={(option, value) => option.title === value.title}
                getOptionLabel={option => option.title}
                options={options}
                loading={loading}
                renderInput={params => (
                    <TextField
                        {...params}
                        label="Nunca dejes de buscar"
                        variant="filled"
                        onChange={ev => {
                            if (ev.target.value !== "" || ev.target.value !== null) {
                                onChangeHandle(ev.target.value);
                            }
                            return ev.target.value
                        }}
                        InputProps={{
                            ...params.InputProps,
                            className: styles.input,
                            disableUnderline: true,
                            endAdornment: (
                                <React.Fragment>
                                    {loading ? (
                                        <CircularProgress color="inherit" size={20} />
                                    ) : null}
                                    {params.InputProps.endAdornment}
                                </React.Fragment>
                            )
                        }}
                    />
                )}
            />
            <Button type="submit" className={styles.searchButton}>
                <div className={styles.searchIcon}></div>
            </Button>
        </form>
    );
}

export default SearchAutocomplete

const ResultInitState: Result[] = []
