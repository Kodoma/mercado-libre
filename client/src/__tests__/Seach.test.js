import React from 'react';
import 'jsdom-global/register';
import {mount, shallow} from 'enzyme';
import Search from '../components/Header/Search/Search';

const useRouter = jest.spyOn(require('next/router'), 'useRouter')

describe('__tests__ Search', () => {
    let component = shallow(<Search />);

    beforeEach(() => {
        component = shallow(<Search />);
    });

    it('should render my Search component', () => {
        expect(component.getElements()).toMatchSnapshot();
    });

    it('should render Search and check legend value', function () {
        useRouter.mockImplementationOnce(() => ({
            query: { search: 'iphone' },
        }))

        const wrap = mount(<Search />)
        expect(wrap.find('form').text()).toBe('Nunca dejes de buscar')
    })

    it('should render Search and check submit button', function () {
        useRouter.mockImplementationOnce(() => ({
            query: { search: 'iphone' },
        }))

        const wrap = mount(<Search />)

        expect(wrap.find('.searchButton')).toBeDefined()

    })

    it('should render Search and check input', function () {
        useRouter.mockImplementationOnce(() => ({
            query: { search: 'iphone' },
        }))

        const wrap = mount(<Search />)

        expect(wrap.find('.input')).toBeDefined()

    })

    it('should submit Search', () => {
        const component = shallow(<Search />);
        let prevented = false;
        component.find('.autocompleteContainer').simulate('submit',{
            preventDefault: () => {
                prevented = true;
            }
        });
        expect(prevented).toBe(true)
    });
});
