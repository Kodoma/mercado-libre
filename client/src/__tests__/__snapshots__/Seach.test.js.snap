// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`__tests__ Search should render my Search component 1`] = `
Array [
  <form
    className="autocompleteContainer"
    onSubmit={[Function]}
  >
    <WithStyles(ForwardRef(Autocomplete))
      freeSolo={true}
      getOptionLabel={[Function]}
      getOptionSelected={[Function]}
      id="q"
      loading={false}
      onClose={[Function]}
      onInputChange={[Function]}
      onOpen={[Function]}
      open={false}
      options={Array []}
      renderInput={[Function]}
      style={
        Object {
          "display": "inline-block",
          "left": "7.5%",
          "marginTop": "8px",
          "position": "relative",
          "width": "70%",
        }
      }
    />
    <WithStyles(ForwardRef(Button))
      className="searchButton"
      type="submit"
    >
      <div
        className="searchIcon"
      />
    </WithStyles(ForwardRef(Button))>
  </form>,
]
`;
